content/blog/2019/06/2019-06-06-cloudfeature-gsoc-cp1.adoc

```diff
@@ -0,0 +1,104 @@
+
+1. For small users, they don't need anybody to manage their jenkins clusters. AWS console gives them a user-friendly to manage the resources, and that would be enough.
+
+2. For medium/large users, it is impossible for them to manage Cloud resources manually (by console). They must use some resource management tools, like CloudFormation by AWS and Terraform by Hashicorp. And me personally are using these tools to manange clusters with thousands of AWS resources. From my perspective, I would like all the resources managed in one place, then that would make me easy to find the problem if something been created/destroyed unexpectedly. I won't let a second participation in my relationship with AWS, as they might potentially ruin everything. e.g, sometimes I had to manage resources using console, but this is even considered as a dangerous operation. Therefore, this plugin should not and not bother to manage any cloud resources. (There are plugins like ec2-plugin who do, but they are well managed, which is safe and used a lot).
+
+For example, when using terraform, we would create an EFS using
+```
+resource "aws_efs_file_system" "efs_jenkins" {
+  encrypted      = false
+  
+	
+  tags = {
+    Name = "Jenkins EFS"
+  }
+}
+```
+Then you would add some mount targets using the same syntax, we skip that here. Finally, you want want to lauch an auto-scaling group or single ec2 instance, and the you write a script like installing the Jenkins in the user data, and all other bootstrap settings.
+
+You have a user data file called ```jenkins-bootstrap.tpl```, like this
+```
+...
+# all dependencied installed
+sudo yum update -y
+sudo yum install jenkins
+sudo systemctl start jenkins.service
+sudo yum install -y amazon-efs-utils
+# then you mount the EFS
+
+sudo mount -t ${EFS_ID}:/ ${MOUNT_POINT}
+```
+Then you create a ec2 instance with such statement
+```
+resource "aws_instance" "jenkins" {
+  ami           = "${data.any_ami_you_want}"
+  instance_type = "t2.micro"
+  user_date     = 
+  tags = {
+    Name = "HelloWorld"
+  }
+}
+
+
+data "template_file" "jenkins_init" {
+  template = "${file("${path.module}/jenkins-bootstrap.tpl")}"
+  vars = {
+    EFS_ID        = "${aws_efs_file_system.efs_jenkins.id}"
+    MOUNT_POINT   = "${var.mount_point}"
+    
+  }
+}
+```
+
+As you can see, user needn't know the efs-id of the created system, just build a dependency like this, and everything is set up automatically. 
+
+3. No convenience is given to user, even we implement the cloud management. We cannot give more feature than CasC(configuration as code) tools and we won't need to. And if we only implement EFS related CasC, what about other resources who also might cooperate with EFS ? e.g. There might not only be one consumer using the EFS, if some service wants to pull information from the Jenkins workspace, that would be inconvenient.
```

Comment:
I think this is where I still don't understand. If the EFS is created by EWS, then EWS owns it and can manage it, including destroying it.